prisma-fmt/src/offsets.rs

```diff
@@ -0,0 +1,134 @@
+/// offsets. This function converts from an LSP position to a pest byte offset. Returns `None` if
+/// the position has a line past the end of the document, or a character position past the end of
+/// the line.
+pub(crate) fn position_to_offset(position: &Position, document: &str) -> Option<usize> {
+    let mut offset = 0;
+    let mut line_offset = position.line;
+    let mut character_offset = position.character as i64;
+    let mut chars = document.chars();
+
+    while line_offset > 0 {
+        loop {
+            match chars.next() {
+                Some('\n') => {
+                    offset += '\n'.len_utf8();
+                    break;
+                }
+                Some(chr) => {
+                    offset += chr.len_utf8();
+                }
+                None => return Some(offset),
+            }
+        }
+
+        line_offset -= 1;
+    }
+
+    while character_offset > 0 {
+        match chars.next() {
+            Some('\n') | None => return Some(offset),
+            Some(chr) => {
+                offset += chr.len_utf8();
+                character_offset -= chr.len_utf16() as i64;
+            }
+        }
+    }
+
+    Some(offset)
+}
+
+#[track_caller]
+/// Converts an LSP range to a span.
+pub(crate) fn range_to_span(range: Range, document: &str, file_id: FileId) -> Span {
+    let start = position_to_offset(&range.start, document).unwrap();
+    let end = position_to_offset(&range.end, document).unwrap();
+
+    Span::new(start, end, file_id)
+}
+
+/// Gives the LSP position right after the given span.
+pub(crate) fn position_after_span(span: Span, document: &str) -> Position {
+    offset_to_position(span.end - 1, document)
+}
+
+/// Converts the byte offset to the offset used in the LSP, which is the number of the UTF-16 code unit.
+pub(crate) fn offset_to_lsp_offset(offset: usize, document: &str) -> usize {
+    let mut current_offset = 0;
+    let mut current_lsp_offset = 0;
+
+    for chr in document.chars() {
+        if offset <= current_offset {
+            break;
+        }
+        current_offset += chr.len_utf8();
+        current_lsp_offset += chr.len_utf16();
+    }
+
+    current_lsp_offset
+}
+
+/// Converts a byte offset to an LSP position, if the given offset
+/// does not overflow the document.
+pub fn offset_to_position(offset: usize, document: &str) -> Position {
+    let mut current_offset = 0;
+    let mut position = Position::default();
+
+    for chr in document.chars() {
+        match chr {
+            _ if offset <= current_offset => {
+                return position;
+            }
+            '\n' => {
+                position.character = 0;
+                position.line += 1;
+            }
+            _ => {
+                position.character += chr.len_utf16() as u32;
+            }
+        }
+        current_offset += chr.len_utf8();
+    }
+
+    position
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Please also add the missing tests added in https://github.com/prisma/prisma-engines/pull/4870 which were also lost because of moving the code